### PR TITLE
WIP start formalizing lms app and launching

### PR DESCRIPTION
**Pull request by jpslav**

### CODE REVIEW

File: app/handlers/lms_complete_launch.rb [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    end
+
+    # Add the user as a teacher or student
+
+    if launch.is_student?
+      # Make sure the user is in the course as a student
+
+      # TODO how do we know which period to add them to?  Possible that we could just add
+      # the student to the course and then on the FE recognize that they are without period
+      # and have them self-select their period?
```
Comment (jpslav) [inline app/handlers/lms_complete_launch.rb]: This is a biggie we need to discuss with UX

File: app/subsystems/lms/launch.rb [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    end
+
+    if (include_everything)
+      data += "\nEverything:\n" + message.ai(plain:true, ruby19_syntax: true, indent: 2)
+    end
+
+    data
+  end
+
+end
```
Comment (nathanstitt) [inline app/subsystems/lms/launch.rb]: are we still only using this in the view?  If so feels like we should move it there and let the view handle the formatting and displaying.

Fine to leave for future though, this is still working towards "real" code

File: db/migrate/20170824032434_reorganize_lms_tables.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+class ReorganizeLmsTables < ActiveRecord::Migration
+
+  # Make sure this class always exist when this migration is run, in case it
+  # goes away one day.
+  class Lms::Models::Nonce < Tutor::SubSystems::BaseModel; end
+
```
Comment (nathanstitt) [inline db/migrate/20170824032434_reorganize_lms_tables.rb]: Great idea - no more legacy migrations that no longer run but aren't discovered until a new team member attempts to start from scratch!

File: db/migrate/20170826122030_create_lms_tables_2.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    end
+
+    create_table :lms_contexts do |t|
+      t.string :lti_id, null: false, index: true
+      t.references :lms_tool_consumer, null: false, index: true,
+                                       foreign_key: { on_update: :cascade, on_delete: :cascade }
+      t.references :course_profile_course, null: false, index: true,
+                                           foreign_key: { on_update: :cascade, on_delete: :cascade }
+    end
+
```
Comment (nathanstitt) [inline db/migrate/20170826122030_create_lms_tables_2.rb]: Pretty sure indexes on `lms_tool_consumers.guid` and `lms_contexts.lti_id` should be unique

File: app/controllers/lms_controller.rb [hunk 5]
```diff
@@ -1,141 +1,85 @@
...
+      launch = Lms::Launch.from_request(request)
 
+      Rails.logger.debug { launch.formatted_data(include_everything: true) }
 
-    @launch_message = authenticator.message
-    # Check that we haven't seen this nonce yet
+      # Bail early if we can't handle the launch role
+      if !(launch.is_student? || launch.is_instructor?)
+        render(:unsupported_role) and return
+      end
```
[the remark this answers is not in the document]
Comment (jpslav) [reply app/controllers/lms_controller.rb]: good point

File: app/subsystems/lms/launch.rb [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  def self.from_id(id)
+    launch_data = Lms::Models::TrustedLaunchData.find(id)
+    raise CouldNotLoadLaunch if launch_data.nil?
+    launch = new(request_parameters: ActiveSupport::HashWithIndifferentAccess.new(launch_data.request_params),
+                 request_url: launch_data.request_url,
+                 trusted: true)
+    launch_data.destroy
+    launch
+  end
+
```
[the remark this answers is not in the document]
Comment (jpslav) [reply app/subsystems/lms/launch.rb]: true - we could leave as a TODO

File: app/subsystems/lms/launch.rb [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+      lms_roles = request_parameters[:roles].split(',')
+      if lms_roles.any?{|lms_role| lms_role.match(/Instructor/)}
+        :instructor
+      elsif lms_roles.any?{|lms_role| lms_role.match(/Student|Learner/)}
+        :student
+      else
+        :other
+      end
+    end
+  end
```
[the remark this answers is not in the document]
Comment (jpslav) [reply app/subsystems/lms/launch.rb]: Yep I think we will want to do those, at least to cover all of the roles in Accounts.

File: app/subsystems/lms/launch.rb (hunk 2 above)
Comment (jpslav) [reply app/subsystems/lms/launch.rb]: I log it to the console only now -- the view is now the real launch

File: db/migrate/20170826122030_create_lms_tables_2.rb (hunk 4 above)
Comment (jpslav) [reply db/migrate/20170826122030_create_lms_tables_2.rb]: combo of guid and lti_id should be unique - that what you're saying?  lti_id _should_ be unique but I don't think it is guaranteed across LMS instances

**Pull request merged**